### ISSUE TITLE
[FR] SCRAM-SHA-1(-PLUS) + SCRAM-SHA-256(-PLUS) + SCRAM-SHA-512(-PLUS) + SCRAM-SHA3-512(-PLUS) supports

### ISSUE THREAD

**Issue by Neustradamus** [docker-mailserver/docker-mailserver#2367]
Dear @docker-mailserver team,

In first, I wish you a Happy New Year!

Can you add supports of :
- SCRAM-SHA-1
- SCRAM-SHA-1-PLUS
- SCRAM-SHA-256
- SCRAM-SHA-256-PLUS
- SCRAM-SHA-512
- SCRAM-SHA-512-PLUS
- SCRAM-SHA3-512
- SCRAM-SHA3-512-PLUS

You can add too:
- SCRAM-SHA-224
- SCRAM-SHA-224-PLUS
- SCRAM-SHA-384
- SCRAM-SHA-384-PLUS

"When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant, and SHA-256 variants [RFC7677] SHOULD be preferred over SHA-1 variants [RFC5802]".

- SCRAM-SHA-1(-PLUS):
-- https://tools.ietf.org/html/rfc5802
-- https://tools.ietf.org/html/rfc6120

- SCRAM-SHA-256(-PLUS):
-- https://tools.ietf.org/html/rfc7677 since 2015-11-02
-- https://tools.ietf.org/html/rfc8600 since 2019-06-21: https://mailarchive.ietf.org/arch/msg/ietf-announce/suJMmeMhuAOmGn_PJYgX5Vm8lNA

- SCRAM-SHA-512(-PLUS):
-- https://tools.ietf.org/html/draft-melnikov-scram-sha-512

- SCRAM-SHA3-512(-PLUS):
-- https://tools.ietf.org/html/draft-melnikov-scram-sha3-512

https://xmpp.org/extensions/inbox/hash-recommendations.html

-PLUS variants:
- RFC5056: On the Use of Channel Bindings to Secure Channels: https://tools.ietf.org/html/rfc5056
- RFC5929: Channel Bindings for TLS: https://tools.ietf.org/html/rfc5929
- Channel-Binding Types: https://www.iana.org/assignments/channel-binding-types/channel-binding-types.xhtml
- Channel Bindings for SCRAM over TLS 1.3: https://tools.ietf.org/html/draft-whited-tls-channel-bindings-for-tls13 -> https://tools.ietf.org/html/draft-ietf-kitten-tls-channel-bindings-for-tls13

LDAP:
- RFC5803: Lightweight Directory Access Protocol (LDAP) Schema for Storing Salted: Challenge Response Authentication Mechanism (SCRAM) Secrets: https://tools.ietf.org/html/rfc5803

HTTP:
- RFC7804: Salted Challenge Response HTTP Authentication Mechanism: https://tools.ietf.org/html/rfc7804

2FA:
- Extensions to Salted Challenge Response (SCRAM) for 2 factor authentication: https://tools.ietf.org/html/draft-melnikov-scram-2fa

IANA:
- Simple Authentication and Security Layer (SASL) Mechanisms: https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml

Linked to:
- https://github.com/scram-xmpp/info/issues/1

**Comment by casperklein** [comment]
We don't implement encryption/hashing on our own. We rely on the implementations in postfix/dovecot themselves.

I don't understand your request for Salted Challenge Response HTTP Authentication Mechanism, as we don't use HTTP at all.. The same goes for 2FA

**Comment by Neustradamus** [comment]
Thanks for your reply!

I have published in the ticket all informations about SCRAM.

Cyrus SASL and Dovecot SASL have supports.

There is a big list (not complete) in "State of Play".

In more, for example:
- PostgreSQL has SCRAM-SHA-256 since 10
- SCRAM-SHA-256-PLUS since 13
- SCRAM-SHA-256 is the default since 14

Another link:
- https://techcommunity.microsoft.com/t5/azure-database-for-postgresql/how-to-securely-authenticate-with-scram-in-postgres-13/ba-p/1548319

A lot of people do not understand that SCRAM-SHA-1 (...) is secure, not like CRAM-MD5 and DIGEST-MD5.
Please read RFCs etc.

Other links:
- https://en.wikipedia.org/wiki/Salted_Challenge_Response_Authentication_Mechanism

**Comment by georglauterbach** [comment]
As @casperklein, we use encryption that Dovecot **and** Postfix already bring with them. What even is the use-case for the algorithms presented here @Neustradamus? Why would we want to use them in DMS? AFAIK the algorithms we use are absolutely fine as is (as of now), so I do not see why we'd need to touch them?

**Comment by Neustradamus** [comment]
I do not see SCRAM here...
- https://github.com/docker-mailserver/docker-mailserver/search?q=md5

**Comment by georglauterbach** [comment]
> I do not see SCRAM here...
> 
>     * https://github.com/docker-mailserver/docker-mailserver/search?q=md5

Yes, it is not used as of now - obviously, otherwise this issue would not exits... But the question I stated two times by now is still open, and I will state it a third time:

**What is the use-case? Why would we want to use these ciphers? Why do the currently used ones not suffice?**

---

If you need them badly, you may alter DMS's configuration with the help of [`user-patches.sh`](https://docker-mailserver.github.io/docker-mailserver/edge/faq/#how-to-adjust-settings-with-the-user-patchessh-script).

**Comment by polarathene** [comment]
> * PostgreSQL has SCRAM-SHA-256 since 10

Why is this mentioned? We do not use Postgres database in this project. Did you confuse it with Postfix? (MTA)

> SCRAM-SHA-256

This is [available as a SASL auth mechanism via Dovecot](https://doc.dovecot.org/configuration_manual/authentication/authentication_mechanisms/#configuration) (we don't use Cyrus anymore). See the linked docs to configure this as supported auth for your mail clients (MUA) via `user-patches.sh` as described above.

> When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant

For this, you'll need to send your request to upstream (Dovecot) to implement, not us.

---

> What is the use-case? Why would we want to use these ciphers? Why do the currently used ones not suffice?

It is not a cipher, but a password auth mechanism (SCRAM) paired with different cryptographic hashes. Main purpose is to send a salted hash to the server instead of only a plaintext password that the server then hashes to verify.

With TLS this assumes the connection has been compromised to read that otherwise encrypted plaintext password, that is where SCRAM _can be_ useful.

Doesn't seem too beneficial:

- SolarDesigner and Sc00bz (_smart crypto people that created Argon2 IIRC, winners of Password Hashing Competition (PHC)_) both chime in [here about SCRAM being not that great](https://www.reddit.com/r/netsec/comments/7xhxd0/the_scram_authentication_protocol/).
- [Another source](https://crypto.stackexchange.com/questions/77286/what-are-the-practical-differences-between-scram-and-pake-opaque-for-simple-au/77288#77288) on SCRAM not likely being much better of a defense against attacks with low entropy inputs (weak passwords).
- Apparently some implementations are poor due to complexity of supporting SCRAM, to the point that [they don't use salted hashed passwords in storage but instead use plaintext](https://github.com/marlam/msmtp-mirror/issues/36#issuecomment-769066588).. Probably not the case with Dovecot of course (_**EDIT:** it actually [can accidentally implicitly store in plaintext](https://doc.dovecot.org/configuration_manual/authentication/password_schemes/#non-plaintext-authentication-mechanisms), be careful!_).
- [An article](https://marlam.de/msmtp/news/about-authentication/) questions the usefulness of SCRAM by reviewing statements from the RFC.

The main benefit is if the secure channel (TLS) is compromised, instead of sending a password in plain text, it's a salted hashed version. But as linked resource above show, SCRAM is not like PAKE / SRP and provides enough information to do an offline attack anyway.

It's a minor hindrance for an attacker who has compromised the TLS connection if the password is weak. If you're using strong passwords, then yes, I can see value in that.

---

I'm not a security expert, but I think that's a fair evaluation.

**TL;DR:** If your concern is with strong passwords being sent in clear text (not hashed) to the server for auth with a potentially compromised TLS connection, SCRAM appears useful.

I don't have time to implement support for configuring that, I'd also advise using `user-patches.sh` approach. Contributing to our documentation about it, or a PR that implements easier configuration support for users is of course welcome.

**Comment by Neustradamus** [comment]
About PostgreSQL is an example, you can see "for example".

I have publised all informations about SCRAM, not specially to you.

Attention, maybe you have not understand and read https://github.com/scram-xmpp/info/issues/1.

It is already done in:
- Cyrus SASL / Cyrus IMAP
- Dovecot
- Exim
- Postfix with Cyrus SASL or Postfix with Dovecot SASL
- msmtp / mpop
- DataEnter CryptoFilter / DataEnter POPBeamer / DataEnter SMTPBeamer / DataEnter XWall
- NeoMutt
- SquirelMail
- SnappyMail
- MimeKit / MailKit
- ...

@jas4711 has added in GNU SASL (gsasl), a long time ago:
- SCRAM-SHA-1 (2009-10-08)
- SCRAM-SHA-1-PLUS (2010-11-14)
- SCRAM-SHA-256/SCRAM-SHA-256-PLUS (2020-01-14)

@marlam has added except -PLUS variants:
- mpop 1.0.20 (SCRAM-SHA-1), and 1.4.13 (SCRAM-SHA-256)
- msmtp 1.4.20 (SCRAM-SHA-1), and 1.8.15 (SCRAM-SHA-256)

TLS Binding 1.3 is in progress, you can see it is in GnuTLS...

I add RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2
- https://tools.ietf.org/html/rfc9051

I wish you a good reading.

**Comment by Neustradamus** [comment]
@georglauterbach, @polarathene: I have added the last RFC in the description: RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2:
- https://tools.ietf.org/html/rfc9051

I wish you a good reading.

**Comment by polarathene** [comment]
> Attention, maybe you have not understand and read [scram-xmpp/info#1](https://github.com/scram-xmpp/info/issues/1).

Understand and read what? A curated link dump on the topic? There's nothing there citing support that differs from what I said for Dovecot, here's the relevant content regarding Dovecot/Postfix:

> > SCRAM-SHA-1 and SCRAM-SHA-256:
> > - Dovecot 2.3.10 (Mail server): https://doc.dovecot.org/configuration_manual/authentication/password_schemes/ + https://dovecot.org/pipermail/dovecot-news/2020-March/000432.html
> > - Postfix with Dovecot SASL: https://postfix.org/
> 
> > SCRAM-SHA-1:
> > - Dovecot: https://doc.dovecot.org/configuration_manual/authentication/password_schemes/ (Dovecot 2.3.10 has SCRAM-SHA-1 and SCRAM-SHA-256)
> > - Postfix with Dovecot SASL: https://postfix.org/

---

> I have publised all informations about SCRAM, not specially to you.
> It is already done in

You are requesting us to provide some implementation / support for SCRAM. You've been directed to upstream Dovecot docs with advice on how you can go about configuring it.

`docker-mailserver` is glue to underlying services, we don't have any involvement with SASL auth beyond delegating to Dovecot. At best we provide some simpler ways to configure via ENV that updates the underlying service configurations.

---

> jas4711 has added in GNU SASL (gsasl), a long time ago
> marlam has added except -PLUS variants
> TLS Binding 1.3 is in progress, you can see it is in GnuTLS...

I don't know what you expect us to do with that information? You probably shouldn't be pining the users here either (_I assume you're doing that elsewhere on the many repo references to your tracking issue, if so that'd be a bit rude spam notifications for those users_).

---

> I have added the last RFC in the description: RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2

Why would you expect us to invest time reading such when you won't even properly engage with us?

[My response to you](https://github.com/docker-mailserver/docker-mailserver/issues/2367#issuecomment-1010483916) is considerably less time on your part than a lengthy RFC document. I wish you a good reading.

**Comment by georglauterbach** [comment]
> Why would you expect us to invest time reading such when you won't even properly engage with us?

Well said - I couldn't have formulated it a better way.
